### ISSUE TITLE
Phase 45.4: CompressionAnalyzer — Compression Analysis & Pareto Optimization

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#892]
### Overview
Implements comprehensive compression analysis framework for evaluating model compression quality across multiple dimensions: parameter reduction, computational savings, memory efficiency, and inference latency. Provides Pareto frontier visualization and automated compression budget allocation.

#### 1. Compression Ratio Measurement
- Parameter count comparison (original vs compressed)
- FLOPs/MACs computation and reduction ratio
- Model file size measurement (on-disk footprint)
- Effective compression ratio: `CR = size_original / size_compressed`
- Bit-level compression analysis for quantized models

#### 2. Accuracy-Efficiency Pareto Frontiers
- Multi-objective evaluation: accuracy vs {latency, memory, params, FLOPs}
- Pareto-optimal configuration identification
- Dominated solution elimination
- Interactive Pareto frontier visualization
- Hypervolume indicator computation for Pareto quality

#### 3. Latency Profiling
- Layer-wise latency measurement on target hardware
- End-to-end inference latency benchmarking
- Throughput measurement (samples/second)
- Hardware simulator integration (CPU, GPU, mobile, edge)
- Roofline model analysis (compute-bound vs memory-bound)
- ONNX Runtime / TensorRT profiling integration

#### 4. Memory Footprint Estimation
- Weight memory: parameter count × bitwidth
- Activation memory: peak activation tensor sizes
- Gradient memory estimation for fine-tuning scenarios
- Runtime memory overhead (buffers, workspace)
- Memory bandwidth utilization analysis

#### 5. Layer-wise Sensitivity Analysis
- Per-layer compression impact on task accuracy
- Sensitivity heatmaps across compression levels
- Critical layer identification (compression-resistant layers)
- Optimal per-layer compression allocation
- Fisher information-based importance scoring

#### 6. Automated Compression Budget Allocation
- Given total compression budget, allocate optimally across layers
- Dynamic programming-based budget allocation
- Gradient-free optimization for budget search
- Hardware-constrained budget allocation

### API Design

```python
class CompressionAnalyzer:
    def __init__(self, config: AnalyzerConfig):
        """Initialize compression analyzer."""
    
    def analyze(
        self,
        original: nn.Module,
        compressed: nn.Module,
        val_loader: DataLoader = None,
        target_hw: str = "cpu"
    ) -> CompressionReport:
        """Generate comprehensive compression analysis report."""
    
    def pareto_frontier(
        self,
        models: List[Tuple[nn.Module, dict]],
        val_loader: DataLoader,
        objectives: List[str] = ["accuracy", "latency"]
    ) -> ParetoFrontier:
        """Compute Pareto frontier across compressed model variants."""
    
    def profile_latency(
        self,
        model: nn.Module,
        input_shape: Tuple[int, ...],
        target_hw: str = "cpu",
        warmup_runs: int = 10,
        benchmark_runs: int = 100
    ) -> LatencyProfile:
        """Profile inference latency on target hardware."""
    
    def sensitivity_analysis(
        self,
        model: nn.Module,
        val_loader: DataLoader,
        compression_type: str = "pruning",
        levels: List[float] = [0.1, 0.3, 0.5, 0.7, 0.9]
    ) -> SensitivityReport:
        """Analyze per-layer compression sensitivity."""
    
    def allocate_budget(
        self,
        model: nn.Module,
        total_budget: float,
        constraint: str = "params",  # params|flops|latency|memory
        val_loader: DataLoader = None
    ) -> BudgetAllocation:
        """Optimally allocate compression budget across layers."""
```

### Acceptance Criteria
- [ ] Parameter, FLOPs, and memory compression ratio computation
- [ ] Pareto frontier visualization with hypervolume indicator
- [ ] Layer-wise latency profiling with hardware targets
- [ ] Memory footprint estimation (weights + activations)
- [ ] Sensitivity heatmap generation
- [ ] Automated compression budget allocation
- [ ] JSON/HTML report generation
- [ ] Unit tests with >90% coverage

### References
- Choudhary et al. (2020) "A Comprehensive Survey on Model Compression and Acceleration"
- Molchanov et al. (2019) "Importance Estimation for Neural Network Pruning"
- Williams et al. (2009) "Roofline: An Insightful Visual Performance Model"
- Zitzler & Thiele (1999) "Multiobjective Evolutionary Algorithms: A Comparative Case Study"

> Part of [Phase 45 Planning](#888)
